File: content/problems/1124-longest-well-performing-interval/solutions/prefix-score-earliest-index/Solution.kt
class Solution {
    fun longestWPI(hours:IntArray):Int{
        val earliest=mutableMapOf<Int,Int>();
        var score=0;
        var answer=0;
        for(i in hours.indices){
            score+=if(hours[i]>8)1 else -1;
            if(score>0)answer=i+1 else if(score-1 in earliest)answer=maxOf(answer,i-earliest[score-1]!!);
            earliest.putIfAbsent(score,i)
        };
        return answer
    }
}

File: content/problems/1124-longest-well-performing-interval/solutions/prefix-score-earliest-index/solution.cpp
class Solution {
    public: int longestWPI(vector<int>& hours){
        unordered_map<int,int> earliest;
        int score=0,answer=0;
        for(int i=0;i<(int)hours.size();++i){
            score+=hours[i]>8?1:-1;
            if(score>0)answer=i+1;
            else if(earliest.count(score-1))answer=max(answer,i-earliest[score-1]);
            if(!earliest.count(score))earliest[score]=i;
        }
        return answer;
    }
};
